rfc/2019-12-07-fuzz-and-auto-debug.md

```diff
@@ -0,0 +1,161 @@
+   |           |execution+----------->+  wrapped app  |                                                |
+   |           +---+-----+      |     +---+-----------+                                                |
+   |  exec result  |            |         ^                                                            |
+   |               v            |         |                                                            |
+   |            +--+---+        |         | get trace info of this input                               |
+   |            |check |        |         |                                                            |
+   |            +--+---+        |         |                                                            |
+   |               |            |         |                                                            |
+   +----------------------------+         |                                                            |
+                   |            +-----------------------------------------------+                      |
+                   |            |         |                        Debug module |                      |
+           input,  |            |         |                                     |                      |
+check pass or not  |            |         |                                     |                      |
+                   |            | +-------+-+       +-----------------+         |                      |
+                   +------------> |collector+------>+statistical model+--------------------------------+
+                                | +---------+       +-------+---------+         |
+                                |                           ^                   |
+                                |                           |                   |
+                                +-----------------------------------------------+
+                                                            |
+                                                            |
+                                                       +----+---+
+                                                       |frontend|
+                                                       +--------+
+
+```
+
+
+
+### Instruction tool
+
+
+
+Like [go tool cover](), we will add code at the beginning of every code basic block in conpiler phase:
+
+```go
+        trace_util_0.Count(_main_00000, 0)   // Instruction
+        flag.Parse()
+        if *version {
+                trace_util_0.Count(_main_00000, 5)  // Instruction
+                fmt.Println(printer.GetTiDBInfo())
+                os.Exit(0)
+        }
+        trace_util_0.Count(_main_00000, 1)   // Instruction
+        registerStores()
+        //.....
+```
+
+`trace_util_0` in above code is an extra package generated by Instruction tool, this package will record start and end line number of every basic block:
+
+```go
+        pathCounter := &CoverCounter{
```

Comment:
ditto

Reply:
OK, I have fixed it

Reply:
The prefix spaces still exists, I'm going to unresolve this conversion.

Reply:
I mistake before. Now I know what you mean.